Refactor ObjectStorage

bolkedebruin opened the pull request:
This refactors the object storage implementation to 
inherit from UPath and therefore from pathlib.Path. 
This makes its behavior semantically stricter and
lowers the maintenance burden.

Furthermore the refactoring includes moving
ObjectStoragePath up a level to `airflow.io.path` making
it more intuitive to import.

Work is underway to allow storage_options to be passed
directly to the underlying filesystem.

cc: @jscheffl @Taragolis @uranusjr 

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

bolkedebruin committed:
Refactor ObjectSToragePath to be a Path

bolkedebruin committed:
Add universal_pathlib

bolkedebruin committed:
Moving stuff to utils

bolkedebruin committed:
Refactor ObjectStoragePath to airflow.io.path

bolkedebruin committed:
Remove redundant parts

bolkedebruin committed:
Update docs

bolkedebruin committed:
Fix mypy and docs

bolkedebruin committed:
Fix errors add tests

bolkedebruin committed:
Some mypy

jscheffl commented on setup.cfg:
```diff
@@ -163,6 +163,7 @@ install_requires =
     # See https://github.com/apache/airflow/pull/31693
     # We should also remove "licenses/LICENSE-unicodecsv.txt" file when we remove this dependency
     unicodecsv>=0.14.1
+    universal_pathlib>=0.1.4
```
I like the clean and lean interface of `upath` looking at the docs. But it is v0.1.4 which sounds like very fesh and instable API. Are you sure it is a good idea to set Airflow fundamental IO basics on a volatile base package? Are there any mature alternatives so that we don't need to feat breaking APIs with every update? (Just mainly asking for stability reasonst, not that I doubt that functional it looks like a nice fit)

bolkedebruin replied:
The nature of `pathlib.Path` I think has prevented alternatives to come to the fore. It did prevent me from deriving from it. I think the risk is limited as `upath` implements the `pathlib.Path` API as a target and therefore will be bound to that.  We could argue we want to pin it specifically to this version to be more predictable. The downside of that is that `upath` might pin a particular dependency (e.g. `fsspec`) in the future (it doesn't now) to something we do not like.

kaxil replied:
>We could argue we want to pin it specifically to this version to be more predictable. The downside of that is that upath might pin a particular dependency (e.g. fsspec) in the future (it doesn't now) to something we do not like.

That's definitely a  big downside

potiuk replied:
> We could argue we want to pin it specifically to this version to be more predictable. The downside of that is that upath might pin a particular dependency (e.g. fsspec) in the future (it doesn't now) to something we do not like.

I think the risk is low and I would not worry about this prematurely.

They (properly like all libraries should) have >= on fsspec. Also we have to remember that there are two dimensions of it. If they currently not pin, but pin in the future, most likely `pip` will resolve it to the current version - not the pinned one if it caused a conflict. As long as there are no security fixes we have to update to, or new features - we are good with staying with the current version (and that one has no upper-binding on fsspec). And fsspec is their only non-dev dependency, so it's really nice and lean library currently.

Also with small libraries like that we can **always** vendor them in - even temporarily, if they are holding us back., We've done that few times. Upath is ~ 1000 lines of code + ~ 40 lines "plugin" per-each-implementation. Some of our PRs are bigger (some of them mine :D ). 

So I would not really worry about what **might** happen - we have options to follow. 

> But it is v0.1.4 which sounds like very fesh and instable API

I don't think that fresh :) . They have 27 (!) releases, 16 contributors and (maybe low but steady stream of contributions - about right when you look at this kind of - small - library. Some of the releases at the beginning were 0.0.* - and likely those were the "fresh". Also I think there is no "universal" notion of 0. being "immature". Just checked and ~150 out of 600 of our dependencies (25%) has 0.* version :).

jscheffl commented on airflow/io/path.py:
```diff
@@ -0,0 +1,396 @@
+        return isinstance(other, _AirflowCloudAccessor) and self._store == other._store
+
+
+class ObjectStoragePath(CloudPath):
```
In my previous comment I was thinking of shortening the name to `StoragePath` but now one step later I realize this is really specific to all sorts of object storage? Means I can not use this IO utility (anymore, compared to previous version?) to also do IO operations on local files with the same API?

bolkedebruin replied:
You can do operations on local files, http (with an `attach`), dbfs (with an `attach`). It just means that such a Path will exhibit the semantics of object storage. This was also the case with the previous implementation. 

We could extend the implementation with a registry - as `upath` does - to support different semantics.  However, dus to the nature of `pathlib.Path` and therefore `upath` it just means repeating quite a lot of code and not being able to fully rely upon upstream implementations. 

In that way `ObjectStoragePath` is more true to its behavior than `StoragePath` would be. An alternative could be `OPath`, but that is also a bit weird I think.

jscheffl reviewed (commented):
First review with a few comments. Mentally an approve but not made full tests. Shall I do a deep testing already now or wait until other comments have settled?

uranusjr committed:
Simplify path accessor

Since attach() already caches store objects, I feel we shouldn't need to
store all the extra information. We can simply take conn_id on Path, and
rely on the internal caching to retrieve the correct information
afterwards.

bolkedebruin committed:
Merge pull request #3 from astronomer/simplify-store-init

Refactor ObjectSToragePath to be a Path

bolkedebruin committed:
Remove need for storing conn_id add experimental

potiuk commented on airflow/io/path.py:
```diff
@@ -0,0 +1,399 @@
+        # remote file -> remote dir
+        if self.is_file():
+            self._cp_file(dst, **kwargs)
+            return
```
NIT: Should we remove if/return? If is not needed here - I believe there is no other option than to have is_file() true here (not is_dir() and exists() means that it is a file). And it feels weird to have both  - if without else and explicit return as the last thing in if.

potiuk replied:
Or maybe just move it to explicit `else:` :)

bolkedebruin replied:
Good catch!

potiuk reviewed (approved):
LGTM - one nit only.

bolkedebruin committed:
Remove redundant if
